### PR TITLE
fix: infinite recursion error by tracking visited nodes

### PR DESCRIPTION
**Pull request by koladilip**
## PR Description

This is reported by bugsnag and it is causing max stack size reached error when circular objects are passed.
Fixes INT-1172

## Linear task (optional)

[Linear task link](https://linear.app/rudderstack/issue/INT-1172/fix-infinite-recursion-error-in-js-sdk-flatten-json-function)

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by mihir-4116 — DISMISSED** [review]

**Review by saikumarrs — DISMISSED** [review]

**Review by mihir-4116 — APPROVED** [review]

File: packages/analytics-js-integrations/__tests__/utils/utils.test.js [hunk 1]
```diff
@@ -303,6 +303,7 @@ describe('flattenJsonPayload Tests', () => {
     const result = utils.flattenJsonPayload(testObj);
     expect(result).toStrictEqual({ '': [] });
   });
+
```
Comment (ItsSudip) [inline packages/analytics-js-integrations/__tests__/utils/utils.test.js:306]:
```suggestion

```

File: packages/analytics-js-integrations/src/utils/utils.js [hunk 2]
```diff
@@ -126,22 +126,42 @@ function rejectArr(obj, fn) {
...
+    for (let i = 0; i < l; i += 1) {
+      recurse(cur[i], prop ? `${prop}.${i}` : `${i}`, res, visited);
+    }
     if (l === 0) res[prop] = [];
-  } else {
+  };
+
+  const processObject = () => {
     let isEmpty = true;
     Object.keys(cur).forEach(key => {
```
Comment (ItsSudip) [inline packages/analytics-js-integrations/src/utils/utils.js:142]: Can't we use for...in here?

Comment (koladilip) [reply packages/analytics-js-integrations/src/utils/utils.js:142]: what would the advantage of it, it is just style right?

Comment (koladilip) [reply packages/analytics-js-integrations/src/utils/utils.js:142]: for..in actually showing a warning to use forEach

**Review by ItsSudip — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**